### PR TITLE
feat: add document presence

### PR DESCRIPTION
**Pull request by hermanwikner**
### Description

This pull request introduces presence in document lists, reference previews and search result items.

**1. Document list**

Screenshot 2022-01-28 at 15 34 57

**2. Reference previews**

1. References auto complete
Screenshot 2022-01-20 at 15 53 24

2. Stand alone reference
Screenshot 2022-01-20 at 15 53 56

3. Array of references auto complete
Screenshot 2022-01-20 at 16 09 55

4. Array of references
Screenshot 2022-01-20 at 15 57 55

**3. Search**

Screenshot 2022-01-20 at 15 49 34

### What to review

- Document status icons (edited and published) and `DocumentPreviewPresence` (when > 2 users) should get the correct color when the item is being `pressed` or `selected`
- Presence should appear as expected in the places mentioned above
- General code review 

### Notes for release

Add presence in document lists, search results and reference previews

### CODE REVIEW

File: packages/@sanity/base/src/datastores/presence/useGlobalDocumentPresence.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import {useMemo} from 'react'
+import {GlobalPresence} from './types'
+import {useGlobalPresence} from './useGlobalPresence'
+
+const EMPTY_ARRAY = []
+
+export function useGlobalDocumentPresence(): Record<string, GlobalPresence[]> {
```
[the remark this answers is not in the document]
Comment (ricokahler) [reply packages/@sanity/base/src/datastores/presence/useGlobalDocumentPresence.tsx:7]: would moving the hook call down a level help?

right now you have this

```js
function Example() {
  const globalPresence = useGlobalDocumentPresence()

  // then...
  const renderItem = useCallback((documentId) => {
    const docPresence = globalPresence[documentId]

    return <Item documentId={documentId} presence={docPresence} />
  }, [globalPresence])

  return // ...
}

function Item({documentId, presence}) {
  return // ...
}
```

is it possible to do this instead using the existing `useDocumentPresence`?

```js
function Example() {
  const renderItem = useCallback((documentId) => {
    return <Item documentId={documentId} />
  }, [])

  return // ...
}

function Item({documentId}) {
  const docPresence = useDocumentPresence(documentId)
}
```

i might be missing some context so i'd double check. @mariuslundgard @bjoerge?

**Review by ricokahler — COMMENTED** [review]
this looks great! i tried it out in the preview and it's working. i did add some comments concerning some perf related items. I'll defer to @bjoerge and @mariuslundgard for some extra eyes.

nice work!

**Review by bjoerge — APPROVED** [review]
Solid work, @hermanwikner!

File: packages/@sanity/base/src/_exports/hooks.ts [hunk 2]
```diff
@@ -1,4 +1,8 @@
-export {useDocumentPresence, useGlobalPresence} from '../datastores/presence/hooks'
+export {
+  useDocumentPresence,
+  useGlobalPresence,
+  useGlobalDocumentPresence,
+} from '../datastores/presence/hooks'
```
Comment (mariuslundgard) [inline packages/@sanity/base/src/_exports/hooks.ts:5]: What is the difference between `useDocumentPresence` and `useGlobalDocumentPresence`?

File: packages/@sanity/base/src/datastores/presence/useGlobalDocumentPresence.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariuslundgard) [reply packages/@sanity/base/src/datastores/presence/useGlobalDocumentPresence.tsx:7]: Yes, I agree with @ricokahler – the API would have a smaller surface, be more performant, and have smoother DX if we solved it like:

```tsx
const documentPresence = useDocumentPresence({documentId})
```

**Review by mariuslundgard — APPROVED** [review]
Looks good, although would be great if we could take a stab at simplifying+refactoring the presence hooks.

File: packages/@sanity/base/src/_exports/hooks.ts (hunk 2 above)
Comment (bjoerge) [reply packages/@sanity/base/src/_exports/hooks.ts:5]: - `useDocumentPresence(id: string)` takes a document id and returns the presence for it
- `useGlobalPresence()` takes no arguments and returns an array of presence items
- `useGlobalDocumentPresence()` is a variant of `useGlobalPresence` that returns a record of `documentId=>PresenceItem[]`.

**Review by ricokahler — APPROVED** [review]
looks great 😄

**Review by bjoerge — APPROVED** [review]
✨

**Pull request merged**